### PR TITLE
Projectfn

### PR DESCRIPTION
**Pull request by yuval-k**

### CODE REVIEW

File: internal/function-discovery/functiontypes/function_types.go [hunk 1]
```diff
@@ -19,6 +20,7 @@ const (
 	FunctionTypeOpenFaas FunctionType = "functionTypeFaas"
 	FunctionTypeAzure    FunctionType = "functionTypeAzure"
 	FunctionTypeFission  FunctionType = "functionTypeFission"
+	FunctionTypeProjectFn  FunctionType = "functionTypeProjectFn"
```
Comment (ilackarms) [inline internal/function-discovery/functiontypes/function_types.go]: gofmt

Comment (ilackarms) [inline internal/function-discovery/functiontypes/function_types.go]: gofmt

File: internal/function-discovery/updater/projectfn/projectfn.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	if us.Type != kubernetes.UpstreamTypeKube {
+		return false
+	}
+
+	spec, err := kubernetes.DecodeUpstreamSpec(us.Spec)
+	if err != nil {
+		return false
+	}
+
+	return strings.HasSuffix(spec.ServiceName, "-fn-api")
```
Comment (ilackarms) [inline internal/function-discovery/updater/projectfn/projectfn.go]: is there perhaps a safer way to determine if an upstream is a project fn upstream?

File: internal/function-discovery/updater/projectfn/projectfn.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	return finalizedroutes, nil
+}
+
+type FnRetreiver struct {
+	client *client.Fn
+}
+
+func createFunction(funk Function) *v1.Function {
```
Comment (ilackarms) [inline internal/function-discovery/updater/projectfn/projectfn.go]: 😂

File: internal/function-discovery/updater/projectfn/projectfn.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+			return appnames, nil
+		}
+	}
+}
+
+func (fr *FnRetreiver) getRoutesForApp(app string) ([]*models.Route, error) {
+
+	var routeobjs []*models.Route
+
+	cursor := ""
```
Comment (ilackarms) [inline internal/function-discovery/updater/projectfn/projectfn.go]: nit: `var cursor string`

File: internal/function-discovery/updater/projectfn/projectfn.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	for {
+		routeParams :=routes.NewGetAppsAppRoutesParams()
```
Comment (ilackarms) [inline internal/function-discovery/updater/projectfn/projectfn.go]: gofmt

File: cmd/function-discovery/main.go [hunk 6]
```diff
@@ -66,6 +66,7 @@ func init() {
 	rootCmd.PersistentFlags().BoolVar(&discoveryOpts.AutoDiscoverGRPC, "detect-grpc-upstreams", true, "enable automatic discovery of upstreams that are running gRPC Services and haeve reflection enabled.")
 	rootCmd.PersistentFlags().BoolVar(&discoveryOpts.AutoDiscoverFaaS, "detect-openfaas", true, "enable automatic discovery for OpenFaaS functions.")
 	rootCmd.PersistentFlags().BoolVar(&discoveryOpts.AutoDiscoverFission, "detect-fission", true, "enable automatic discovery for Fission functions.")
+	rootCmd.PersistentFlags().BoolVar(&discoveryOpts.AutoDiscoverProjectFn, "detect-project-fn-upstreams", true, "enable automatic discovery project fn upstreams.")
```
Comment (ilackarms) [inline cmd/function-discovery/main.go]: can we rename this `detect-project-fn`?

File: internal/function-discovery/updater/projectfn/imported/swagger.yml [hunk 7]
```diff
@@ -0,0 +1,259 @@
+swagger: '2.0'
```
Comment (ilackarms) [inline internal/function-discovery/updater/projectfn/imported/swagger.yml]: suggest adding a comment here describing the file?

File: internal/function-discovery/updater/projectfn/projectfn.go (hunk 2 above)
Comment (ilackarms) [reply internal/function-discovery/updater/projectfn/projectfn.go]: if not this seems fine as an early appproach

File: internal/function-discovery/updater/projectfn/projectfn.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	return strings.HasSuffix(spec.ServiceName, "-fn-api")
+}
+
+func GetFuncs(resolve resolver.Resolver, us *v1.Upstream) ([]*v1.Function, error) {
+	return GetFuncsWithTransport(resolve, us, nil)
+}
+
+func GetFuncsWithTransport(resolve resolver.Resolver, us *v1.Upstream, transport runtime.ClientTransport) ([]*v1.Function, error) {
```
[the remark this answers is not in the document]
Comment (yuval-k) [reply internal/function-discovery/updater/projectfn/projectfn.go]: no, as I use it in a test

File: internal/function-discovery/updater/projectfn/projectfn.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+		return nil, err
+	}
+	var finalizedroutes []*v1.Function
+	for _, funk := range funcs {
+		finalizedroutes = append(finalizedroutes, createFunction(funk))
+	}
+	return finalizedroutes, nil
+}
+
+func (fr *FnRetreiver) GetFuncs() ([]Function, error) {
```
[the remark this answers is not in the document]
Comment (yuval-k) [reply internal/function-discovery/updater/projectfn/projectfn.go]: it used in a test

File: internal/function-discovery/updater/projectfn/imported/swagger.yml (hunk 7 above)
Comment (yuval-k) [reply internal/function-discovery/updater/projectfn/imported/swagger.yml]: I prefer not to change this file, as it is imported form projectfn. I'll add some comments to gen.go

**Comment by yuval-k** [comment]
@ilackarms  PTAL

File: internal/function-discovery/updater/projectfn/projectfn.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ilackarms) [reply internal/function-discovery/updater/projectfn/projectfn.go]: what about putting the test in the same package?

**Pull request merged**